assets/src/vo/money.js

```diff
@@ -0,0 +1,519 @@
+	 * Rounds towards nearest neighbour. If equidistant, rounds towards
+	 * Infinity.
+	 * @type {number}
+	 */
+	static ROUND_HALF_CEIL = Decimal.ROUND_HALF_CEIL;
+
+	/**
+	 * Rounds towards nearest neighbour. If equidistant, rounds towards
+	 * -Infinity.
+	 * @type {number}
+	 */
+	static ROUND_HALF_FLOOR = Decimal.ROUND_HALF_FLOOR;
+
+	/**
+	 * Class constructor
+	 * @param {number|string|Decimal} amount
+	 * @param {Currency} currency
+	 */
+	constructor( amount, currency ) {
+		this.setCurrency( currency )
+			.setAmount( amount )
+			.setFormatter();
+		Object.freeze( this );
+	}
+
+	/**
+	 * Set the currency property
+	 *
+	 * @param {Currency} currency
+	 * @return {Money} Either this Money or new Money depending on state of
+	 * property.
+	 */
+	setCurrency( currency ) {
+		Money.assertCurrency( currency );
+		// if there's already a currency set, then return a new object.
+		if ( this.currency instanceof Currency ) {
+			return new Money( this.amount, currency );
+		}
+		this.currency = currency;
+		return this;
+	}
+
+	/**
+	 * Set the amount property
+	 *
+	 * @param {Decimal|number|string} amount
+	 * @return {Money} Either this Money or new Money depending on state of the
+	 * property.
+	 */
+	setAmount( amount ) {
+		const value = amount instanceof Decimal ? amount.toNumber() : amount;
+		// if there's already an amount set, then return a new object.
+		if ( this.amount instanceof Decimal ) {
+			return new Money( new Decimal( value ), this.currency );
+		}
+		this.amount = new Decimal( value );
+		return this;
+	}
+
+	/**
+	 * Set the formatter for money values
+	 *
+	 * @return {Money} An instance of this object.
+	 */
+	setFormatter() {
+		// only initialize if its not already initialized
+		if ( isEmpty( this.formatter ) ) {
+			Accounting.settings = {
+				...Accounting.settings,
+				...this.currency.toAccountingSettings(),
+			};
+			this.formatter = Accounting;
+		}
+		return this;
+	}
+
+	/**
+	 * Returns the value of this Money as its subunits.
+	 * @return {number} If the subunits is 100 and the value is .45,
+	 * this returns 450
+	 */
+	toSubunits() {
+		return this.amount.toNumber() * this.currency.subunits;
+	}
+
+	/**
+	 * Returns whether the provided money object equals this money object.
+	 * Compares both amount and currency.
+	 *
+	 * @param {Money} other
+	 * @return {boolean} True means this is equal. False means it isn't.
+	 */
+	equals( other ) {
+		Money.assertMoney( other );
+		return this.amount.equals( other.amount ) &&
+			this.hasEqualCurrency( other );
+	}
+
+	/**
+	 * Returns whether provided Money object's Currency equals this Money
+	 * object's Currency.
+	 *
+	 * This does a shallow comparison on the serialized values for the currency
+	 * objects.  That way if the currencies are different instances, but share
+	 * the same internal value, they are considered equal.
+	 *
+	 * @param {Money} other
+	 * @return {boolean} True means the currencies are equal.
+	 */
+	hasEqualCurrency( other ) {
```

Comment:
> using hasEqual sounds like they could actually be different but still somehow have equality in some other sense.

This is exactly why it is this way.  This allows for them to be different _instances_ of an object but still equal in _value_. If someone needs to know whether two instances are identical they can just do a `===` check.  The purpose of this method is to determine equivalency of value (via a shallow compare of the value represented by the currency).  You can view the tests to see the implication of this.

> ie: currencies WXY and XYZ are both trading at a value of 1.25 relative to currency ABC and can therefore be considered equivalent, despite not being the same currency.

This example doesn't really apply here because in that case this method would return `WXY` and `XYZ` as being non-equal.  Again, please look at the tests and it will become clearer how this works.  If after the tests it still isn't clear ping me in chat.

Reply:
Also when I talk about value here remember we're talking about the _currency_ NOT the _money_ instance.  So _value_ represents the configuration for that currency.

Reply:
> please look at the tests and it will become clearer how this works.

I realize exactly how it works and what is actually happening in the method but my issue is that the naming potentially **sounds** like something else is happening. To be clear, I'm somewhat splitting hairs because the terms "equal" and "same" in many cases are interchangeable, however, in this context I don't think they are, and using "equal" potentially conveys something else, whereas using "same" would be unequivocally clear.

Remember that with Value Objects we don't care about instances. We only care about values. So two instances with equal values are not just considered "equal", they are considered to be the **same** thing, which is exactly why I'm suggesting the name change. Anybody doing instance checks with Value Objects is doing it wrong because one of the fundamental cornerstones of the pattern is that individual instances do not matter. An object representing USD is the **same** as any other object representing USD. They don't just have equal values, we literally consider them to be the same thing, so they are completely interchangeable. Two things with "equal" values should not be considered value objects if we are considering them to be separate things, but should instead be considered entities because we are considering their identity. Two Money objects that both use a USD Currency object are considered to be sharing the same Currency, even though it is almost guaranteed that the actual instances are different (common with VOs since new instances are constantly being created). It's a subtle but important difference. Remember, we never ever ever care about instances, so two Money objects with USD Currency are sharing the same Currency object. 

If we converted `hasEqualCurrency()` into a plain sentence/question, it might be something like:

- does this money have an equal currency?
- is the currency for this other amount of money equal to this amount of money?
- are these two money amounts displayed with an equal currency?
- do these two money amounts have an equal currency?

First thing right off the bat... people don't talk like that.
It sounds totally awkward and weird, so why use weird awkward phrasing to name a method?

Secondly, using the term "equal" leads to the question "equal in what way?". 
Using the term "equal" means that we are indeed comparing values. You even stated that outright in your response:

>  The purpose of this method is to determine equivalency of value 

and my beef with that, is that it's not inherently clear whether we doing that for ALL of the values, or only some of them. Using the term "same" makes it clear that ALL of the values are equal and not just a subset.

Thirdly, the above sentences sound like we are potentially talking about more than one currency (ex: do these two money amounts have an equal currency?),  whereas using "same" offers a slight change in meaning and makes it clear that there is only **one** currency in question:

- does this money have the same currency?
- is the currency for this other amount of money the same as this amount of money?
- are these two money amounts displayed with the same currency?
- do these two money amounts have the same currency?

Here's another way to approach this...
let's put both terms into one sentence and separate them with an "OR" to determine which one better conveys the concept of to things being identical:

### is this an equal currency or is this the same currency?

one of those terms unequivocally conveys the idea that we are talking about something that is identical in every way (remember we don't care about instances).

bottom line: `hasEqualCurrency()` is just not as unequivocally clear as `hasSameCurrency()`

Reply:
It's clear from the thread so far that we both think the choice of term matters. Essentially if I summarize the conflicting positions here:

### My argument

Using the term "same" (within code) is equivalent to `===` or strict equality because it refers to something that is identical (not only the same value, but also the same type or instance).  Using the term "equals" (in code) is equivalent to `==` or loose equality because it refers to something that has the same value (but may not be the same type or instance).

**Since its true we don't care about whether the instances are the same** in value objects and all we care about is whether the value is equal, I chose to use the term equals.  I think for the purposes of the methods described by this term it is clear what it is doing.

### Your Argument

`hasEqualCurrency()` is just not as unequivocally clear as `hasSameCurrency()` because you believe that `hasSameCurrency()` conveys the idea that we are talking about something that is _identical in every way_ (because we don't care about instances).  

Essentially the disagreement (as I see it) boils down to you saying that the meaning of "same" (or identical) in the context of vo's is different than in any other code context because things that might disqualify something from being the same outside a vo are not applicable to inside a vo.  Whereas my point is that I don't want to change the meaning of a term to fit the context, but rather use the meaning of a term that is _still useful_ in this context.  

Put another way.  When comparing value objects, it doesn't matter if they are completely identical (equal in instance AND value), what matters is they are _equal_ in value only.  So I chose the term "equals".  

As for your plain english examples, while in some cases trying to sound out how we'd talk about code in plain English can help clarify things, I think it also can be misleading because English is not a very precise language.  In every day conversation, currency is more of an abstract concept.  People use currency to describe the kind of money they have, "I get paid in US Dollars, not Canadian dollars".  So in _everyday conversation_ the concept of "same" or "equal" applies to the dollar as a unit.  I'm not going to say "Does this dollar have an equal (or same) currency as that dollar?"  No, we're more likely to say "The Canadian dollar is not equal in value (or the same in value) as the US dollar".  In everyday conversation there's essentially no difference between usage of the terms "same" and "equals".

However when we get to writing code, the choice of terms has more meaning, and that's why I think it matters what we choose to describe the behaviour of a method.   I _know_ instances don't matter in value objects, so I'm not going to use the term **same** for a method that compares the _values_ of two instances.   Instead I chose what I feel is the _still useful_ and more accurate - the term **equals**. Functionally that's all we care about.

Reply:
> Using the term "same" (within code) is equivalent to === or strict equality because it refers to something that is identical (not only the same value, but also the same type or instance). 

exactly... and since we are talking about VOs you absolutely have to drop the concept of instances because they don't exist with VOs. If all of the values are the same, then they are identical. It's true that you can create multiple instances of a VO **in code** (which is in fact a really common occurrence with VOs since that's how you maintain immutability), but you should never ever ever consider those instances in any way whatsoever. You basically have to ignore their existence. Why? Because context totally matters when you are attempting to model something from the real world. Just because something may be true with regards to the code you write, does not make it true in the real world. There is only one USD currency. Trying to conceptualize multiple USD currencies makes as much sense as saying something is 1 km north of the North pole. You can't have 1 instance of USD currency in one hand and another instance of USD currency in the other, at least not without bringing in the concept of money itself. It doesn't make sense in the real world. You also can't add the USD Currency to itself (USD + USD = how many USD !?!?), but we could do it in code right? We could create thousands of instances of USD Currency in code, but in the real world, there is only ever going to be **one** USD Currency, and that's what we are trying to model. USD Currency is USD Currency is USD Currency. They are all identical to any other USD Currency, ie: 

```
USD Currency === USD Currency
```

two amounts of USD money share the same currency. We don't say their currencies are equal because that implies there is more than one currency in play. ie: "the currency for this $10 USD is equal to the currency for this other $10 USD". Thinking like that doesn't make sense.

> Using the term "equals" (in code) is equivalent to == or loose equality because it refers to something that has the same value 

but loose equality is not necessarily the same value depending on the **context** of what you are referring to. Look at PHP's "truthy" values:

```php
1 == "1" == true // all loosely equal
0 == "0" == false == null // all loosely equal
```

sure those values are certainly interchangeable in a lot of places (if not most)... but NOT everywhere... the fact is... context DOES matter.

> (but may not be the same type or instance). 

**This is exactly my point!!!**

The term "equals" can potential mean that two things (and this is right from your own wording): **may not be the same**
And again, because we are talking about a currency VO, we have to completely ignore instances (because, you know, real world), so conveying that something may potentially not be equal (while ignoring instances) means that one of the values **may not be the same**.

This is why I'm being so insistent about using "same" vs "equal" because with regards to currency, we don't want to convey that things are "loosely equivalent, but may not be the same" we want to convey that they are identical.

> my point is that I don't want to change the meaning of a term to fit the context

but again, context totally matters.

Here's something that may (or may not) blow your mind (or at least make you go "huh")... for our purposes, money can be conceptualized as a value object, because again, we don't care about instances, and only care about amounts and what currency those amounts are in. Nearly all stores and businesses fall into this category. If however, we were writing software for a central bank (or the Federal Reserve in the US) then we would want to conceptualize money as an Entity and suddenly the concept of identity would become an important property. At this point though we would probably be modeling "Notes", or "Bills", and would use the serial numbers printed on physical money for the individual identifiers. This is one of the main conceptual tenants of Domain Driven Design, which is that your data schema is modeled entirely around the domain, which is the real world space being emulated, and the meanings of all terms depend on the context they are used within.

So in one domain, Money can be modeled as a value object with no identity and the concept of equality depends solely on having matching amounts and currency, and in another domain, Money needs to be modeled as an entity where each item has a separate identity, and the concept of identical equality likely infers the presence of a federal crime (because no two bills should have the exact same serial number). Context totally matters and the concept of equality changes depending on the context.

> As for your plain english examples, while in some cases trying to sound out how we'd talk about code in plain English can help clarify things, I think it also can be misleading because English is not a very precise language. In every day conversation, currency is more of an abstract concept. People use currency to describe the kind of money they have, "I get paid in US Dollars, not Canadian dollars". So in everyday conversation the concept of "same" or "equal" applies to the dollar as a unit. I'm not going to say "Does this dollar have an equal (or same) currency as that dollar?" No, we're more likely to say "The Canadian dollar is not equal in value (or the same in value) as the US dollar". In everyday conversation there's essentially no difference between usage of the terms "same" and "equals".

mmmmm... ya not sure what your point is since you kinda switched over to talking about money and/or exchange rates and not currency on its own. 
If we stick to talking strictly about currency as a separate concept that describes money, then we CAN easily imagine an everyday conversation that uses it within that context:

> store clerk: Should I put this $10 bill in with all of the other $10 bills?
> manager: If it's the same currency, then ya of course. If it's not the same currency, then put it under the money tray.

Now say that last sentence but use "equal" instead:

> manager: If it's an equal currency, then ya of course. If it's not an equal currency, then put it under the money tray.

ya... nobody is going to talk like that, although to be fair they would probably same something like:

> manager: If it's (reference to country) $10, then ya of course. If it's not (reference to country), then put it under the money tray.

So with regards to everyday conversation and/or plain language usage nobody is going to favour the term "equals" over "same" when talking about a currency. With regards to the technical aspect of things, the only way that "equals" makes more sense is if we introduce the concept of instances because then the inference of something being identical that comes with the term "same" wouldn't be applicable to the Currency VO. But that is totally doing it wrong because you should not consider instances when dealing with value objects because they do not have identity. The argument for this is basically saying:

> the term equals is more accurate... because if we were referring to entities with separate identities and the individual instances were important... then it would make sense!

but we are not referring to entities with separate identities, we're referring to value objects and the individual instances should be disregarded!

This argument is equivalent to:

> Why are you driving on the wrong side of the road?
> If we were in a country where everyone drove on this side of the road then it would make sense to drive on this side
> ...uhhhh... but we're not in one of those countries...

Same issue here. If we were talking about entities, then instances and identities, and the concepts of equality and sameness would take on different meanings because the context would be different. That's not the case here. Similarly, if we were attempting to model an entirely different domain, then we would be dealing with different contexts and may have to approach familiar concepts from an entirely different angle.

Anyways, I'm sorry for being a pain and being so stubborn about this but I think naming is important and want to ensure that things have the most intuitive names they can have. Once this gets out there it will be hard to change so let's get it perfect now. I think I'll post something in chat to get some feedback from the rest of the team with regards to what they feel would be a more intuitive naming schema.

Reply:
> I think I'll post something in chat to get some feedback from the rest of the team with regards to what they feel would be a more intuitive naming schema.

A Grinder post might be better for discussion.